c++/src/kj/async-test.c++

```diff
@@ -699,6 +699,47 @@ TEST(Async, ArrayJoinVoid) {
   promise.wait(waitScope);
 }
 
+TEST(Async, ArrayJoinFailFast) {
+  EventLoop loop;
+  WaitScope waitScope(loop);
+
+  constexpr static auto ELEMENT_COUNT = 5;
+
+  auto builder = heapArrayBuilder<Promise<void>>(ELEMENT_COUNT);
+  Vector<Own<PromiseFulfiller<void>>> fulfillers;
+
+  uint tasksCompleted = 0;
+
+  for (auto KJ_UNUSED value: zeroTo(ELEMENT_COUNT)) {
+    auto paf = newPromiseAndFulfiller<void>();
+    builder.add(paf.promise.then([&tasksCompleted]() {
+      ++tasksCompleted;
+    }));
+    fulfillers.add(kj::mv(paf.fulfiller));
+  }
+
+  Promise<void> promise = joinPromisesFailFast(builder.finish());
+
+  {
+    uint i = 0;
+    KJ_EXPECT(tasksCompleted == 0);
+
+    // Joined tasks are completed eagerly, not waiting until the join node is awaited.
+    fulfillers[i++]->fulfill();
+    KJ_EXPECT(!promise.poll(waitScope));
+    KJ_EXPECT(tasksCompleted == i);
+
+    fulfillers[i++]->fulfill();
+    KJ_EXPECT(!promise.poll(waitScope));
+    KJ_EXPECT(tasksCompleted == i);
+
+    fulfillers[i++]->reject(KJ_EXCEPTION(FAILED, "Test exception"));
+
+    // The first rejection makes the promise ready.
+    KJ_EXPECT_THROW_MESSAGE("Test exception", promise.wait(waitScope));
+  }
+}
+
```

Comment:
Might be worth creating the same test but for `joinPromises` instead, both to show their differences clearly but also to ensure coverage.

Reply:
I agree; I made a version of this test for `joinPromises()`, and also refactored the existing `joinPromises()` happy-path tests to run against `joinPromisesFailFast()`.